benchmark/httpapi.py

```diff
@@ -0,0 +1,215 @@
+
+from klein import Klein
+
+from zope.interface import implementer
+
+from ._interfaces import IBackend
+
+
+class ResultNotFound(Exception):
+    """
+    Exception indicating that a result with a given identifier is not found.
+    """
+
+
+@implementer(IBackend)
+class DummyBackend(object):
+    """
+    The backend that simply drops all results.
+    """
+    def store(self, result):
+        """
+        Pretend to store a single benchmarking result.
+
+        :param dict result: The result in the JSON compatible format.
+        :return: A Deferred that produces an identifier for the stored
+            result.
+        """
+        return succeed(uuid4().hex)
+
+    def retrieve(self, id):
+        """
+        Just fail.
+        """
+        return fail(ResultNotFound())
+
+    def query(self, filter):
+        """
+        Return an empty result.
+        """
+        return succeed(list())
+
+    def delete(self, id):
+        """
+        Just fail.
+        """
+        return fail(ResultNotFound())
+
+
+@implementer(IBackend)
+class InMemoryBackend(object):
+    """
+    The backend that simply drops all results.
+
+    :ivar dict results: Stored results by their identifiers.
+    """
+    def __init__(self):
+        self.results = {}
+
+    def store(self, result):
+        """
+        Store a single benchmarking result and return its identifier.
+
+        :param dict result: The result in the JSON compatible format.
+        :return: A Deferred that produces an identifier for the stored
+            result.
+        """
+        id = uuid4().hex
+        self.results[id] = result
+        return succeed(id)
+
+    def retrieve(self, id):
+        """
+        Retrive a result by the given identifier.
+        """
+        try:
+            return self.results[id]
+        except KeyError:
+            return fail(ResultNotFound())
+
+    def query(self, filter):
+        """
+        Return matching results.
+        """
+        matching = [
+            r for r in self.results.viewvalues()
+            if filter.viewitems() <= r.viewitems()
+        ]
+        return succeed(matching)
+
+    def delete(self, id):
+        """
+        Delete a result by the given identifier.
+        """
+        try:
+            del self.results[id]
+        except KeyError:
+            return fail(ResultNotFound())
+
+
+class BenchmarkAPI_V1(object):
+    """
+    API for storing and accessing benchmarking results.
+
+    :ivar IBackend backend: The backend for storing the results.
+    """
+    app = Klein()
+    version = 1
+
+    def __init__(self, backend):
+        """
+        :param IBackend backend: The backend for storing the results.
+        """
+        self.backend = backend
+
+    @app.handle_errors(ResultNotFound)
+    def _not_found(self, request, failure):
+        request.setResponseCode(404)
+        return ""
+
+    @app.route("/submit", methods=['POST'])
+    def submit(self, request):
```

Comment:
Apologies if I'm covering ground that's already been covered. 

I think we would generally prefer to have a more RESTful API. That would mean having an endpoint for `/benchmark-results` (or similar) and then POSTing to that when we want to add a new one, rather than having a dedicated endpoint for submitting.